front: ngetoosrd: fix and streamline handle create and update

Synar opened the pull request:
Close [#13506](https://github.com/OpenRailAssociation/osrd/issues/13506)

Depend on https://github.com/OpenRailAssociation/netzgrafik-editor-frontend/pull/687 (bump commit will be added as soon as that pr or one fixing the same issue is merged) -> pr merged, still waiting for the fork to be bumped

~~The code is a bit cumbersome at the moment. It should become much simpler as soon as paced trains and train schedules get merged. I propose we delay further refactoring until that point.~~ Done

Also close [#14766](https://github.com/OpenRailAssociation/osrd/issues/14766). Both changes were done in the same pr to ease testing and avoid conflicts.

Math-R commented on front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd/trainrun.ts:
```diff
@@ -436,41 +433,135 @@ const handleCreateTimetableItem = async (
+      ]);
+      addUpsertedTimetableItems([newForwardTrainSchedule, newReturnTrainSchedule]);
+    }
   }
```
My though is that the nested if/else for handling one_way/round_trip and paced/non-paced creates 4 nearly identical branches.
The only things that change are:

- Which endpoint to call (PacedTrains vs TrainSchedules)
- Which ID formatter to use
- Body size ([trip] vs [trip, returnTrip])
- RoundTrips structure (one_ways vs round_trips)
- Expected item count (1 vs 2)

Maybe you should calculated these parameters in first place and then use a single code path, what do you think about this ?

Synar replied:
Indeed, there's a lot of repeated code here. The main issue with refactoring is typing : paced train and train schedules are incompatible, and so we would need to have ifs at every step in order to pass the type checking (even though we know there won't be issues).

This is why I proposed delaying refactoring this code until paced trains and train schedules are merged, as this hurdle will go away.

I also think it would be nice to refactor the branching paths of update at the same time.

clarani reviewed (commented):
Code LGTM, but I won't have the time to test today :/ I can test tomorrow !

Yohh commented on front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd/trainrun.ts:
```diff
@@ -436,41 +433,135 @@ const handleCreateTimetableItem = async (
+
+  const paced = createPacedAttributesFromTrainrun(trainrun, netzgrafikDto);
+
+  if (trainrun.direction === 'one_way') {
```
in this if..else bloc, a lot of things are repeated, I'm not sure to have the time to test anything before monday but I'm pretty sure we can improve this to make conditions in a single bloc, more readable in the end

Yohh replied:
anyway I read the upper comment, sorry I didn't check it before, you can merge if you you have an approve

Synar commented:
Well now that the train schedules / paced trains merger got merged, I suppose I can refactor this pr now ^^

Synar commented:
Done! This version is much simpler and shorter, thanks to the aforementioned merging. It can still be simplified further once round trip endpoints get merged (assuming they will).

clarani commented on front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/ngeToOsrd/trainrun.ts:
```diff
@@ -435,41 +431,54 @@ const handleCreateTimetableItem = async (
+
+  const postRoundTripsEndpoint = paced
+    ? osrdEditoastApi.endpoints.postRoundTripsPacedTrains
+    : osrdEditoastApi.endpoints.postRoundTripsTrainSchedules;
```
Hum I think `postRoundTripsTrainSchedules` is going to be dropped, you should always use `postRoundTripsPacedTrains`

Synar replied:
But we're still querying from the train schedule endpoint currently, right? Shouldn't we do the switch in an atomic way?

Synar replied:
But we're still querying from the train schedule endpoint currently, right? Shouldn't we do the switch in an atomic way?

Synar replied:
Although I suppose it really does not matter much

Synar replied:
Although I suppose it really does not matter much

Synar replied:
These endpoints were dropped in another pr, so done!

clarani reviewed (commented):
Codes LGTM :white_check_mark:
I will test once it's rebase.

Synar committed:
front: let createPacedTrain create multiple trains at once

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar committed:
front: ngetoosrd: type the return of createPacedAttr

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar committed:
front: ngetoosrd: fix train duplication

Trains without paced and one way trains were not supported by handleCreateTimetableItem

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar committed:
front: ngetoosrd: fix and streamline handleUpdate

handleUpdateTimetableItem was broken after the paced and train schedule merger,
and its structure was overly complex with duplicated code (which previously served to treat paced and train schedules differently)

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar committed:
front: bump nge

Signed-off-by: Alice Khoudli <alice.khoudli@polytechnique.org>

Synar commented:
This keeps getting new conflicts, so I would prefer we don't wait too long to merge it

Math-R reviewed (approved):
looks good to me :)
